### PR TITLE
[BEAM-1320] Add script to generate documentation for the python sdk

### PR DESCRIPTION
**Pull request by sb2nov**
R: @aaltay 

- Created a script to generate the documentation for pydocs to be checked into the beam site.
- Added sphinx as a dev dependency
- There will be a corresponding PR in the beam-site repo

---

Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [ ] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [ ] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [ ] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [ ] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -49,6 +49,9 @@ hs_err_pid*.log
 # Ignore MacOSX files.
 .DS_Store
 
+# Ignore the documentation build files in pydocs
+sdks/python/docs
+
 # NOTE: if you modify this file, you probably need to modify the file set that
```
Comment (davorbonaci) [inline .gitignore]: please take a peek at this note; can easily break source distribution / rat plugin.

File: sdks/python/setup.py [hunk 2]
```diff
@@ -101,6 +101,7 @@ def get_version():
 
 REQUIRED_TEST_PACKAGES = [
     'pyhamcrest>=1.9,<2.0',
+    'Sphinx>=1.5.2,<2.0'
```
Comment (davorbonaci) [inline sdks/python/setup.py]: comments? why `test`? alternatives?

File: sdks/python/generate_pydocs.sh [hunk 3]
```diff
@@ -0,0 +1,79 @@
+#!/bin/bash
```
Comment (davorbonaci) [inline sdks/python/generate_pydocs.sh]: bind the script to the maven `package` lifecycle to make sure it doesn't break and can simplify the instructions, and make it platform-specific. (See what archetypes do.)

File: sdks/python/generate_pydocs.sh [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+#
+# Use "generate_pydocs.sh" to update documentation in the docs directory.
+#
+# The exit-code of the script indicates success or a failure.
+
+# Quit on any errors
+set -e
+
+# Create docs directory if it does not exist
+mkdir -p docs/source
```
Comment (davorbonaci) [inline sdks/python/generate_pydocs.sh]: can you do `target/docs/source` instead?

File: sdks/python/generate_pydocs.sh [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+]
+master_doc = 'index'
+html_theme = 'sphinxdoc'
+project = 'Apache Beam'
+EOF
+cat > docs/source/index.rst <<'EOF'
+Welcome to Apache Beam!
+=======================
+
+`Apache Beam <http://beam.apache.org>`_ is a unified model for defining both batch and streaming data-parallel processing pipelines. Beam provides a set of language-specific SDKs for constructing pipelines. These pipelines can be executed on distributed processing backends like `Apache Spark <http://spark.apache.org/>`_, `Apache Flink <http://flink.apache.org/>`_, and `Google Cloud Dataflow <http://cloud.google.com/dataflow>`_.
```
Comment (davorbonaci) [inline sdks/python/generate_pydocs.sh]: https://beam.apache.org/

File: sdks/python/generate_pydocs.sh [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+
+Please see the `Apache Beam website <http://beam.apache.org>`_ for more documention around usage and contributing.
```
Comment (davorbonaci) [inline sdks/python/generate_pydocs.sh]: likewise

File: sdks/python/setup.py (hunk 2 above)
Comment (sb2nov) [inline sdks/python/setup.py]: We can add it as an extra_require too, will update. I added it here that it was used only in dev.

File: sdks/python/generate_pydocs.sh (hunk 4 above)
Comment (sb2nov) [inline sdks/python/generate_pydocs.sh]: do you mean `root/target/` or `sdks/python/target` ? Is there a reason for this?

Comment (davorbonaci) [inline sdks/python/generate_pydocs.sh]: `sdks/python/target`

The reason is that `mvn clean` will remove it. Rat plugin will ignore it. Assembly plugin for source release will skip it.

It would be awesome if *all* temporary files across Python SDK could be in `target/`. It would simplify some things a lot.

File: .gitignore (hunk 1 above)
Comment (sb2nov) [inline .gitignore]: Removed

File: sdks/python/setup.py (hunk 2 above)
Comment (sb2nov) [inline sdks/python/setup.py]: Moved to extras

Comment (sb2nov) [inline sdks/python/setup.py]: I'm not sure how jenkins runs this. We might just want to install it as part of requirements.

File: sdks/python/generate_pydocs.sh (hunk 3 above)
Comment (sb2nov) [inline sdks/python/generate_pydocs.sh]: Done

File: sdks/python/generate_pydocs.sh (hunk 4 above)
Comment (sb2nov) [inline sdks/python/generate_pydocs.sh]: DOne

File: sdks/python/generate_pydocs.sh (hunk 5 above)
Comment (sb2nov) [inline sdks/python/generate_pydocs.sh]: Removed the page as it wasn't need.

File: sdks/python/generate_pydocs.sh (hunk 6 above)
Comment (sb2nov) [inline sdks/python/generate_pydocs.sh]: Removed the page as it wasn't need.

File: sdks/python/setup.py [hunk 7]
```diff
@@ -103,6 +103,10 @@ def get_version():
     'pyhamcrest>=1.9,<2.0',
     ]
 
+EXTRA_REQUIRES = {
```
Comment (davorbonaci) [inline sdks/python/setup.py]: but does this pushes a requirement to users or developers (or both)?

File: sdks/python/pom.xml [hunk 8]
```diff
@@ -39,131 +39,165 @@
...
-              </environmentVariables>
-            </configuration>
-          </execution>
-        </executions>
-      </plugin>
-    </plugins>
-  </build>
+  <profiles>
+    <profile>
+      <id>beam-sdks-python-package</id>
```
Comment (davorbonaci) [inline sdks/python/pom.xml]: pushing everything into a profile looks unrelated to this PR. Can you revert?

File: sdks/python/pom.xml [hunk 9]
```diff
@@ -39,131 +39,165 @@
...
+                </configuration>
+              </execution>
+            </executions>
+          </plugin>
+        </plugins>
+      </build>
+    </profile>
+    <profile>
+      <id>beam-sdks-python-pydoc</id>
+      <activation>
```
Comment (davorbonaci) [inline sdks/python/pom.xml]: should be auto-activated by the release plugin and/or part of the existing release profile

File: sdks/python/setup.py (hunk 7 above)
Comment (sb2nov) [inline sdks/python/setup.py]: Only for developers to install this.

File: sdks/python/pom.xml (hunk 9 above)
Comment (sb2nov) [inline sdks/python/pom.xml]: Removed the default false

File: sdks/python/pom.xml (hunk 8 above)
Comment (sb2nov) [inline sdks/python/pom.xml]: Activation based on OS can only be set in a profile, I'm not very familiar with maven but was following

[1] https://github.com/apache/beam/blob/master/sdks/java/maven-archetypes/pom.xml#L60
[2] http://stackoverflow.com/questions/8963713/maven-condition-based-on-os-family

Comment (davorbonaci) [inline sdks/python/pom.xml]: We have a misunderstanding -- let's reset.

There are 2 profiles:
* Regular profile that runs most of Python - Maven integration. AFAIK, that should work on any OS and shouldn't be in a profile. Always on, no profiles -- exactly as is today.
* The comment about the second profile is below.

File: sdks/python/pom.xml (hunk 9 above)
Comment (davorbonaci) [inline sdks/python/pom.xml]: We have a misunderstanding -- let's reset.

There are 2 profiles:
* The comment about the first profile is above.
* The comment about Pydoc one is here.

My understanding:
* Runs on Linux-like OS only, because it requires `bash`. (Separately, converting to a python script might be better, since it presumably wouldn't be platform specific.)
* Should be in a profile that is _not_ on by default, exactly as Javadoc is, so that build time is not negatively affected for development scenarios.
* The profile should be on for pre-commit coverage, i.e., if Pydoc stops building, pre-commit should fail.
* The profile should be automatically activated when running a release (by virtue of the release plugin and release profile).

Happy to help with any of this, as you see fit ;-)

Comment (sb2nov) [inline sdks/python/pom.xml]: Thanks for explaining. I'll try to do this and will let you know if I need more help with maven.

**Comment by sb2nov** [comment]
@davorbonaci I created https://issues.apache.org/jira/browse/BEAM-1509 as the python pom.xml is broken in multiple places example check https://github.com/apache/beam/blob/master/sdks/python/pom.xml#L133. 

The current PR should get the documentation to be part of pre-commits and I'll work on fixing the pom file in another PR. Is that okay ?

**Comment by sb2nov** [comment]
@davorbonaci ping?
